src/radical/pilot/agent/radical-pilot-agent-multicore.py

```diff
@@ -1170,6 +1179,190 @@ def corner2offset(self, block, corner):
+            return True
+        else:
+            return False
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _release_slot(self, opaque_slot):
+        #-----------------------------------------------------------------------
+        # One application has finished, increase the number of available slots.
+        #with self._slot_lock:
+        self._log.info('Releasing : {0} Cores, {1} RAM'.format(opaque_slot['task_slots'][0],opaque_slot['task_slots'][1]))
+        self.avail_cores +=opaque_slot['task_slots'][0]
+        self.avail_mem +=opaque_slot['task_slots'][1]
+        self.avail_app['apps']+=1
+        return True
+
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _try_allocation(self, cu):  
+        """
+        Attempt to allocate cores for a specific CU.  If it succeeds, send the
+        CU off to the ExecutionWorker.
+        """
+        #-----------------------------------------------------------------------
+        # Check if the YARN scheduler queue has space to accept new CUs.
+        # Check about racing conditions in the case that you allowed an
+        # application to start executing and before the statistics in yarn have
+        # refreshed, to send another one that does not fit.
+
+        # TODO: Allocation should be based on the minimum memor allocation per
+        # container. Each YARN application needs two containers, one for the 
+        # Application Master and one for the Container that will run.
+        
+        # needs to be locked as we try to acquire slots, but slots are freed
+        # in a different thread.  But we keep the lock duration short...
+        with self._slot_lock :        
+
+            self._log.info(self.slot_status())
+            self._log.debug('YARN Service and RM URLs: {0} - {1}'.format(self._service_url,self._rm_url))
+                    
+            cu['opaque_slots']={'lm_info':{'service_url':self._service_url,
+                                            'rm_url':self._rm_url,
+                                            'nodename':self._client_node},
+                                'task_slots':[cu['description']['cores'],2048]
+                                            }
+
+            alloc = self._allocate_slot(cu['description']['cores'],2048)
+
+        if not alloc:
+            return False
+
+        # got an allocation, go off and launch the process
+        self._prof.prof('schedule', msg="allocated", uid=cu['_id'])
+        self._log.info("slot status after allocated  : %s" % self.slot_status ())
+
+        return True
+
+    # --------------------------------------------------------------------------
+    #
+    def work(self, cu):
+
+      # self.advance(cu, rp.AGENT_SCHEDULING, publish=True, push=False)
+        self._log.info("Overiding Parent's class method")
+        self.advance(cu, rp.ALLOCATING      , publish=True, push=False)
```

Comment:
Who would have though I'd ever say this: there are too many spaces!  You may want to distribute them elsewhere... ;)